File: test/stat_test.cpp
/* Copyright (c) 2024-2025. The FSMOD Team. All rights reserved.          */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

#include <gtest/gtest.h>
#include <iostream>

#include <simgrid/s4u/Engine.hpp>
#include <simgrid/s4u/Actor.hpp>

#include "fsmod/FileSystem.hpp"
#include "fsmod/OneDiskStorage.hpp"
#include "fsmod/FileSystemException.hpp"

#include "./test_util.hpp"

namespace sgfs=simgrid::fsmod;
namespace sg4=simgrid::s4u;

XBT_LOG_NEW_DEFAULT_CATEGORY(stat_test, "Stat Test");

class StatTest : public ::testing::Test {
public:
    std::shared_ptr<sgfs::FileSystem> fs_;
    sg4::Host * host_;
    sg4::Disk * disk_;

    StatTest() = default;

    void setup_platform() {
        XBT_INFO("Creating a platform with one host and two disks...");
        auto *my_zone = sg4::Engine::get_instance()->get_netzone_root()->add_netzone_full("zone");
        host_ = my_zone->add_host("my_host", "100Gf");
        disk_ = host_->add_disk("disk", "1kBps", "2kBps");
        my_zone->seal();

        XBT_INFO("Creating a one-disk storage on the host's disk...");
        auto ods = sgfs::OneDiskStorage::create("my_storage", disk_);
        XBT_INFO("Creating a file system...");
        fs_ = sgfs::FileSystem::create("my_fs");
        XBT_INFO("Mounting a 1MB partition...");
        fs_->mount_partition("/dev/a/", ods, "1MB");
    }
};

TEST_F(StatTest, Stat)  {
    DO_TEST_WITH_FORK([this]() {
        this->setup_platform();
        // Create one actor (for this test we could likely do it all in the maestro but what the hell)
        host_->add_actor("TestActor", [this]() {
            std::shared_ptr<sgfs::File> file;
            XBT_INFO("Create a 100kB file at /dev/a/foo.txt");
            ASSERT_NO_THROW(fs_->create_file("/dev/a/foo.txt", "100kB"));
            XBT_INFO("Open File '/dev/a/foo.txt'");
            ASSERT_NO_THROW(file = fs_->open("/dev/a/foo.txt", "r"));
            XBT_INFO("Calling stat()");
            std::unique_ptr<sgfs::FileStat> stat_struct;
            ASSERT_NO_THROW(stat_struct = file->stat());
            XBT_INFO("Checking sanity");
            ASSERT_EQ(stat_struct->size_in_bytes, 100*1000);
            ASSERT_DOUBLE_EQ(stat_struct->last_modification_date, simgrid_get_clock());
            ASSERT_DOUBLE_EQ(stat_struct->last_access_date, simgrid_get_clock());
            ASSERT_EQ(stat_struct->refcount, 1);
            XBT_INFO("Modifying file state");
            std::shared_ptr<sgfs::File> file2;
            ASSERT_NO_THROW(file2 = fs_->open("/dev/a/foo.txt", "w"));
            ASSERT_NO_THROW(file2->seek(100*1000));
            ASSERT_NO_THROW(file2->write(12*1000, false));
            ASSERT_NO_THROW(file->close());
            ASSERT_NO_THROW(sg4::this_actor::sleep_for(10));
            ASSERT_NO_THROW(file2 = fs_->open("/dev/a/foo.txt", "r"));
            ASSERT_NO_THROW(file->seek(0));
            ASSERT_NO_THROW(file->read(1000, false));
            XBT_INFO("Calling stat() again");
            ASSERT_NO_THROW(stat_struct = file->stat());
            XBT_INFO("Checking sanity");
            ASSERT_EQ(stat_struct->size_in_bytes, 112*1000);
            ASSERT_DOUBLE_EQ(stat_struct->last_modification_date, simgrid_get_clock() - 10);
            ASSERT_DOUBLE_EQ(stat_struct->last_access_date, simgrid_get_clock());
            ASSERT_EQ(stat_struct->refcount, 2);
            XBT_INFO("Modifying file state");
            ASSERT_NO_THROW(file2->close());
            XBT_INFO("Calling stat() again");
            ASSERT_NO_THROW(stat_struct = file->stat());
            XBT_INFO("Checking sanity");
            ASSERT_EQ(stat_struct->refcount, 1);
            ASSERT_NO_THROW(file->close());
        });

        // Run the simulation
        ASSERT_NO_THROW(sg4::Engine::get_instance()->run());
    });
}

File: test/python/stat_test.py
import sys
import multiprocessing
from simgrid import Engine, this_actor, ActivitySet, StorageFailureException
from fsmod import FileSystem, OneDiskStorage, File, NotEnoughSpaceException

def setup_platform():
    e = Engine(sys.argv)
    e.set_log_control("no_loc")
    e.set_log_control("root.thresh:critical")

    # Creating a platform with one host and one disk...
    zone = e.netzone_root.add_netzone_full("zone")
    host = zone.add_host("my_host", "100Gf")
    disk = host.add_disk("disk", "1kBps", "2kBps")
    zone.seal()

    # Creating a one-disk storage on the host's disk..."
    ods = OneDiskStorage.create("my_storage", disk)
    # Creating a file system
    fs = FileSystem.create("my_fs")
    # Mounting a 100MB partition
    fs.mount_partition("/dev/a/", ods, "1MB")

    return e, host, fs

def run_test_stat():
    e, host, fs = setup_platform()
    def test_actor():
        this_actor.info("Create a 100kB file at /dev/a/foo.txt")
        fs.create_file("/dev/a/foo.txt", "100kB")
        this_actor.info("Open File '/dev/a/foo.txt'")
        file = fs.open("/dev/a/foo.txt", "r")
        this_actor.info("Calling stat()")
        stat_struct = file.stat()
        this_actor.info("Checking sanity")
        assert stat_struct.size_in_bytes == 100*1000
        assert stat_struct.last_modification_date == Engine.clock
        assert stat_struct.last_access_date == Engine.clock
        assert stat_struct.refcount == 1
        this_actor.info("Modifying file state")
        file2 = fs.open("/dev/a/foo.txt", "w")
        file2.seek(100*1000)
        file2.write(12*1000, False)
        file.close()
        this_actor.sleep_for(10)
        file2 = fs.open("/dev/a/foo.txt", "r")
        file.seek(0)
        file.read(1000, False)
        this_actor.info("Calling stat() again")
        stat_struct = file.stat()
        this_actor.info("Checking sanity")
        assert stat_struct.size_in_bytes == 112*1000
        assert stat_struct.last_modification_date== Engine.clock - 10
        assert stat_struct.last_access_date== Engine.clock
        assert stat_struct.refcount == 2
        this_actor.info("Modifying file state")
        file2.close()
        this_actor.info("Calling stat() again")
        stat_struct = file.stat()
        this_actor.info("Checking sanity")
        assert stat_struct.refcount == 1
        file.close()

    host.add_actor("TestActor", test_actor)
    e.run()
  
if __name__ == '__main__':
    tests = [
      run_test_stat
    ]

    for test in tests:
        print(f"\n🔧 Run {test.__name__} ...")
        p = multiprocessing.Process(target=test)
        p.start()
        p.join()

        if p.exitcode != 0:
           print(f"❌ {test.__name__} failed with exit code {p.exitcode}")
        else:
           print(f"✅ {test.__name__} passed")
